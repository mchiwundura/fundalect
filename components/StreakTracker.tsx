import React, { useEffect, useState } from "react";
import { View, Text, StyleSheet } from "react-native";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { ThemedText } from "./ThemedText";
import ThemedView from "./ThemedView";
import { StreakData } from "@/types/ui/streakTracker";
import { areConsecutive, getDateNumber, getDayLabel, getMidWeekDates, isToday } from "@/utils/streak";



export default function StreakTracker() {
  const STORAGE_KEY = "streakData";
  const [streakData, setStreakData] = useState<StreakData>({
    currentStreak: 0,
    bestStreak: 0,
    activeDays: [],
  });


  useEffect(() => {
    const loadData = async () => {
      try {
        const json = await AsyncStorage.getItem(STORAGE_KEY);
        if (json) {
          const data: StreakData = JSON.parse(json);
          const sortedDays = data.activeDays.sort(
            (a, b) => new Date(a.date).getTime() - new Date(b.date).getTime()
          );
          setStreakData({ ...data, activeDays: sortedDays });
        }
      } catch (error) {
        console.error("Failed to load streak data", error);
      }
    };

    loadData();
  }, []);

  const { currentStreak, bestStreak, activeDays } = streakData;

  const completedDates = new Set(activeDays.map((entry) => entry.date));
  const displayDates = getMidWeekDates();

  return (
    <ThemedView style={styles.container}>
      <Text style={styles.streakText}>ðŸ”¥ {currentStreak} Days</Text>
      <Text style={styles.bestStreakText}>Best: {bestStreak} Days</Text>

      <View style={styles.weekContainer}>
        {displayDates.map((dateStr, index) => {
          const isCompleted = completedDates.has(dateStr);
          const prevDate = displayDates[index - 1];
          const nextDate = displayDates[index + 1];

          const isPrevConnected =
            isCompleted && completedDates.has(prevDate) && areConsecutive(prevDate, dateStr);
          const isNextConnected =
            isCompleted && completedDates.has(nextDate) && areConsecutive(dateStr, nextDate);

          return (
            <View key={dateStr} style={styles.dayItem}>
              <ThemedView
                style={[

  styles.dayBlock,
  isCompleted ? styles.completed : styles.incomplete,

  // Fully connected streak block (middle)
  isCompleted && isPrevConnected && isNextConnected && styles.fullySquare,

  // Connected to previous only
  isCompleted && !isPrevConnected && isNextConnected && styles.leftRounded,

  // Connected to next only
  isCompleted && isPrevConnected && !isNextConnected && styles.rightRounded,

  // Standalone completed day
  isCompleted && !isPrevConnected && !isNextConnected && styles.fullyRounded,


                ]}
              >
                <ThemedText style={[styles.dayText, isToday(dateStr) && styles.todayText, isCompleted && styles.completedDayText]}>
                  {getDateNumber(dateStr)}
                </ThemedText>
              </ThemedView>
              <ThemedText type="default">{getDayLabel(dateStr)}</ThemedText>
            </View>
          );
        })}
      </View>
    </ThemedView>
  );
}

const styles = StyleSheet.create({
  container: {
    padding: 20,
    borderRadius: 15,
    height: 150
  },
  streakText: {
    fontSize: 16,
    fontWeight: "bold",
    color: "orange",
  },
  bestStreakText: {
    fontSize: 14,
    color: "gray",
  },
  weekContainer: {
    flexDirection: "row",
    justifyContent: "center",
    marginTop: 10,
    flexWrap: "nowrap",
    width: "100%",
  },
  dayItem: {
    marginHorizontal: 3,
    position: "relative",
    minWidth: 30,
    height: 55,
    paddingTop: 35,
    display: "flex",
    alignItems: "center",
  },
  dayBlock: {
    paddingHorizontal: 6,
    top: 0,
    left: 0,
    right: 0,
    height: 30,
    position: "absolute",
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    boxShadow: 'none'
  },
  completed: {
    backgroundColor: "#9C85FF",
  },
  incomplete: {
    borderWidth: 1,
    borderColor: "#9C85FF",
    borderRadius: 15,
  },
  leftRounded: {
    borderTopLeftRadius: 15,
    borderBottomLeftRadius: 15,
    width: 36,
    marginRight: -6,
  },
  rightRounded: {
    borderTopRightRadius: 15,
    borderBottomRightRadius: 15,
    width: 33,
    marginLeft: -3,
  },
  fullyRounded: {
    borderRadius: 15,
  },
  fullySquare: {
    borderRadius: 0,
    width: 36,
    marginHorizontal: 0,
  },
  dayText: {
    fontWeight: "bold",
  },
  todayText: {
    color: "#FFD700", 
  },
  completedDayText: {
    color: 'rgba(255, 255, 255, 0.8)',
  }
});
