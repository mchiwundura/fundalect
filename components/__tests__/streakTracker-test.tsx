import React from "react";
import { render, waitFor } from "@testing-library/react-native";
import StreakTracker from "../StreakTracker";
import AsyncStorage from "@react-native-async-storage/async-storage";

// Mock AsyncStorage
jest.mock("@react-native-async-storage/async-storage", () => ({
  getItem: jest.fn(),
  setItem: jest.fn(),
}));

describe("StreakTracker snapshot tests", () => {
  const mockData = {
    currentStreak: 3,
    bestStreak: 7,
    activeDays: [
      { date: "2025-06-01" },
      { date: "2025-06-02" },
      { date: "2025-06-03" },
    ],
  };

  beforeEach(() => {
    (AsyncStorage.getItem as jest.Mock).mockResolvedValue(
      JSON.stringify(mockData),
    );
  });

  it("renders correctly and matches snapshot", async () => {
    let tree;

    tree = render(<StreakTracker />);

    await waitFor(() => {
      expect(tree.toJSON()).toMatchSnapshot();
    });
  });
});
