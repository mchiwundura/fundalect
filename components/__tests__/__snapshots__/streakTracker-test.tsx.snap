// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StreakTracker snapshot tests renders correctly and matches snapshot 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "rgba(255, 255, 255, 0.7)",
        "boxShadow": "0 4px 4px rgba(0, 0, 0, 0.25)",
      },
      {
        "borderRadius": 15,
        "height": 150,
        "padding": 20,
      },
    ]
  }
>
  <Text
    style={
      {
        "color": "orange",
        "fontSize": 16,
        "fontWeight": "bold",
      }
    }
  >
    ðŸ”¥ 
    0
     Days
  </Text>
  <Text
    style={
      {
        "color": "gray",
        "fontSize": 14,
      }
    }
  >
    Best: 
    0
     Days
  </Text>
  <View
    style={
      {
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "justifyContent": "center",
        "marginTop": 10,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "height": 55,
          "marginHorizontal": 3,
          "minWidth": 30,
          "paddingTop": 35,
          "position": "relative",
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "rgba(255, 255, 255, 0.7)",
              "boxShadow": "0 4px 4px rgba(0, 0, 0, 0.25)",
            },
            [
              {
                "alignItems": "center",
                "boxShadow": "none",
                "display": "flex",
                "height": 30,
                "justifyContent": "center",
                "left": 0,
                "paddingHorizontal": 6,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              {
                "borderColor": "#9C85FF",
                "borderRadius": 15,
                "borderWidth": 1,
              },
              false,
              false,
              false,
              false,
            ],
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgba(0,0,0,0.8)",
              },
              {
                "fontSize": 16,
                "lineHeight": 24,
              },
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              [
                {
                  "fontWeight": "bold",
                },
                false,
                false,
              ],
            ]
          }
        >
          4
        </Text>
      </View>
      <Text
        style={
          [
            {
              "color": "rgba(0,0,0,0.8)",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        W
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "height": 55,
          "marginHorizontal": 3,
          "minWidth": 30,
          "paddingTop": 35,
          "position": "relative",
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "rgba(255, 255, 255, 0.7)",
              "boxShadow": "0 4px 4px rgba(0, 0, 0, 0.25)",
            },
            [
              {
                "alignItems": "center",
                "boxShadow": "none",
                "display": "flex",
                "height": 30,
                "justifyContent": "center",
                "left": 0,
                "paddingHorizontal": 6,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              {
                "borderColor": "#9C85FF",
                "borderRadius": 15,
                "borderWidth": 1,
              },
              false,
              false,
              false,
              false,
            ],
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgba(0,0,0,0.8)",
              },
              {
                "fontSize": 16,
                "lineHeight": 24,
              },
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              [
                {
                  "fontWeight": "bold",
                },
                false,
                false,
              ],
            ]
          }
        >
          5
        </Text>
      </View>
      <Text
        style={
          [
            {
              "color": "rgba(0,0,0,0.8)",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        T
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "height": 55,
          "marginHorizontal": 3,
          "minWidth": 30,
          "paddingTop": 35,
          "position": "relative",
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "rgba(255, 255, 255, 0.7)",
              "boxShadow": "0 4px 4px rgba(0, 0, 0, 0.25)",
            },
            [
              {
                "alignItems": "center",
                "boxShadow": "none",
                "display": "flex",
                "height": 30,
                "justifyContent": "center",
                "left": 0,
                "paddingHorizontal": 6,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              {
                "borderColor": "#9C85FF",
                "borderRadius": 15,
                "borderWidth": 1,
              },
              false,
              false,
              false,
              false,
            ],
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgba(0,0,0,0.8)",
              },
              {
                "fontSize": 16,
                "lineHeight": 24,
              },
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              [
                {
                  "fontWeight": "bold",
                },
                false,
                false,
              ],
            ]
          }
        >
          6
        </Text>
      </View>
      <Text
        style={
          [
            {
              "color": "rgba(0,0,0,0.8)",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        F
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "height": 55,
          "marginHorizontal": 3,
          "minWidth": 30,
          "paddingTop": 35,
          "position": "relative",
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "rgba(255, 255, 255, 0.7)",
              "boxShadow": "0 4px 4px rgba(0, 0, 0, 0.25)",
            },
            [
              {
                "alignItems": "center",
                "boxShadow": "none",
                "display": "flex",
                "height": 30,
                "justifyContent": "center",
                "left": 0,
                "paddingHorizontal": 6,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              {
                "borderColor": "#9C85FF",
                "borderRadius": 15,
                "borderWidth": 1,
              },
              false,
              false,
              false,
              false,
            ],
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgba(0,0,0,0.8)",
              },
              {
                "fontSize": 16,
                "lineHeight": 24,
              },
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              [
                {
                  "fontWeight": "bold",
                },
                {
                  "color": "#FFD700",
                },
                false,
              ],
            ]
          }
        >
          7
        </Text>
      </View>
      <Text
        style={
          [
            {
              "color": "rgba(0,0,0,0.8)",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        S
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "height": 55,
          "marginHorizontal": 3,
          "minWidth": 30,
          "paddingTop": 35,
          "position": "relative",
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "rgba(255, 255, 255, 0.7)",
              "boxShadow": "0 4px 4px rgba(0, 0, 0, 0.25)",
            },
            [
              {
                "alignItems": "center",
                "boxShadow": "none",
                "display": "flex",
                "height": 30,
                "justifyContent": "center",
                "left": 0,
                "paddingHorizontal": 6,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              {
                "borderColor": "#9C85FF",
                "borderRadius": 15,
                "borderWidth": 1,
              },
              false,
              false,
              false,
              false,
            ],
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgba(0,0,0,0.8)",
              },
              {
                "fontSize": 16,
                "lineHeight": 24,
              },
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              [
                {
                  "fontWeight": "bold",
                },
                false,
                false,
              ],
            ]
          }
        >
          8
        </Text>
      </View>
      <Text
        style={
          [
            {
              "color": "rgba(0,0,0,0.8)",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        S
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "height": 55,
          "marginHorizontal": 3,
          "minWidth": 30,
          "paddingTop": 35,
          "position": "relative",
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "rgba(255, 255, 255, 0.7)",
              "boxShadow": "0 4px 4px rgba(0, 0, 0, 0.25)",
            },
            [
              {
                "alignItems": "center",
                "boxShadow": "none",
                "display": "flex",
                "height": 30,
                "justifyContent": "center",
                "left": 0,
                "paddingHorizontal": 6,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              {
                "borderColor": "#9C85FF",
                "borderRadius": 15,
                "borderWidth": 1,
              },
              false,
              false,
              false,
              false,
            ],
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgba(0,0,0,0.8)",
              },
              {
                "fontSize": 16,
                "lineHeight": 24,
              },
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              [
                {
                  "fontWeight": "bold",
                },
                false,
                false,
              ],
            ]
          }
        >
          9
        </Text>
      </View>
      <Text
        style={
          [
            {
              "color": "rgba(0,0,0,0.8)",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        M
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "height": 55,
          "marginHorizontal": 3,
          "minWidth": 30,
          "paddingTop": 35,
          "position": "relative",
        }
      }
    >
      <View
        style={
          [
            {
              "backgroundColor": "rgba(255, 255, 255, 0.7)",
              "boxShadow": "0 4px 4px rgba(0, 0, 0, 0.25)",
            },
            [
              {
                "alignItems": "center",
                "boxShadow": "none",
                "display": "flex",
                "height": 30,
                "justifyContent": "center",
                "left": 0,
                "paddingHorizontal": 6,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              {
                "borderColor": "#9C85FF",
                "borderRadius": 15,
                "borderWidth": 1,
              },
              false,
              false,
              false,
              false,
            ],
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "rgba(0,0,0,0.8)",
              },
              {
                "fontSize": 16,
                "lineHeight": 24,
              },
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
              [
                {
                  "fontWeight": "bold",
                },
                false,
                false,
              ],
            ]
          }
        >
          10
        </Text>
      </View>
      <Text
        style={
          [
            {
              "color": "rgba(0,0,0,0.8)",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        T
      </Text>
    </View>
  </View>
</View>
`;
